File: WebUI/script/types/CtrRef.js
class CtrRef {
    constructor(typeName, name,  partitionGuid, instanceGuid) {
        this.typeName = typeName;
        this.name = name;
        this.partitionGuid = partitionGuid;
        this.instanceGuid = instanceGuid;
    }

    setFromTable(table) {
        this.typeName = table.typeName;
        this.name = table.name;
        this.partitionGuid = table.partitionGuid;
        this.instanceGuid = table.instanceGuid;
    }

    clone() {
        return new CtrRef(this.typeName, this.name, this.partitionGuid, this.instanceGuid);
    }
}

File: ext/Shared/Types/CtrRef.lua
---@class CtrRef
CtrRef = class 'CtrRef'

local m_Logger = Logger("CtrRef", false)

function CtrRef:__init(arg)
	---@type string
	self.typeName = arg.typeName
	---@type string
	self.name = arg.name
	---@type Guid
	self.partitionGuid = arg.partitionGuid
	---@type Guid
	self.instanceGuid = arg.instanceGuid
end

function CtrRef:GetTable()
	return {
		typeName = self.typeName,
		name = self.name,
		partitionGuid = tostring(self.partitionGuid),
		instanceGuid = tostring(self.instanceGuid)
	}
end

function CtrRef:Get()
	local s_Instance = ResourceManager:FindInstanceByGuid(Guid(self.partitionGuid), Guid(self.instanceGuid))

	-- TODO: More resolving for custom objects or whatever? Idk
	if s_Instance then
		s_Instance = _G[s_Instance.typeInfo.name](s_Instance)
		return s_Instance
	end

	return nil
end

return CtrRef
